### DISCUSSION TITLE
Is the data secure?

### DISCUSSION THREAD

**Discussion by keiron1992** [livewire/livewire#4388] [1. Help] [answered]
Hi all,

Please forgive my lack of understanding with this, I'm grateful for anyone taking the time to respond.

If I pass the following to a view from a controller

`
$public = $data

///Livewire Controller
public function render(){
    $this->data = ['username' => 'bob', 'password' => 'psswd', 'phone' => '12345678'];
    return view('test', $this->data)
}

//Lievwire view
<div>
   {{$username}}
</div>
`

Am I correct is saying that despite me passing the variables to the view, only the ones echo'd are visible to the user and that the remaining data is not passed to the browser in any way?

For example the password or phone number is not accessible to the user?

Is that correct?

Thanks,
Keiron

**Comment by jonneroelofs** [comment, marked as answer]
Anything that you define as a public property on a livewire component gets send between the browser and the server on every request an so is visible to users if they were for example to look at the network requests. If you take a look at network requests in your dev tools and open a livewire request and checkout the payload, then you can see all the data.

I can highly recommend this blog post by the author of Livewire, Caleb Porzio:
https://calebporzio.com/how-livewire-works-a-deep-dive

**Comment by joshhanley** [comment]
@keiron1992 as @jonneroelofs said above any public properties data are sent to the front end.

One way you can make that more secure is to just pass your data directly into the view instead of assigning it to `$this->data` but then that data won't be reactive for Livewire.

Like:

```php
public function render(){
return view('test', ['username' => 'bob', 'password' => 'psswd', 'phone' => '12345678']);
}
```

Hope this helps!

**Reply by Jul000** [reply]
Hi, Thanks for your answer. Do you know if user can see the public function in this case ? for example if we make database query in a public function ?

Thanks,

Julien

**Reply by Padrezz** [reply]
When you click on livewire div, you can see which data it's sending with requests. So, as i know, it's sending just function names.

**Reply by joshhanley** [reply]
@Julien0510 no, the public function isn't sent to the front end, so you can safely make a db query in it.

**Reply by Jul000** [reply]
Alright thank you

**Reply by joshhanley** [reply]
No worries! 🙂
